cassiopeia/simulator/BirthDeathFitnessSimulator.py

```diff
@@ -0,0 +1,336 @@
+                        ] += (min_total_time - remaining_lineage["total_time"])
+                        tree.nodes[remaining_lineage["id"]][
+                            "birth_scale"
+                        ] = tree.nodes[parent]["birth_scale"]
+                    break
+            # If extant tips are the stopping criteria, pop the minimum age
+            # lineage at each step
+            if num_extant:
+                min_age_ind = np.argmin(
+                    [i["total_time"] for i in current_lineages]
+                )
+                lineage = current_lineages.pop(min_age_ind)
+            else:
+                lineage = current_lineages.pop(0)
+            for _ in range(2):
+                unique_id = sample_event(unique_id, lineage)
+
+        # Prune dead lineages and collapse resulting unifurcations
+        if death_waiting_dist and len(tree.nodes) > 1:
+            if experiment_time:
+                for i in list(tree.nodes):
+                    if (
+                        tree.out_degree(i) == 0
+                        and tree.nodes[i]["total_time"] < experiment_time
+                    ):
+                        self.remove_and_prune_lineage(i, tree)
+            if num_extant:
+                surviving_ids = [i["id"] for i in current_lineages]
+                for i in list(tree.nodes):
+                    if tree.out_degree(i) == 0 and i not in surviving_ids:
+                        self.remove_and_prune_lineage(i, tree)
+            if len(tree.nodes) > 1:
+                self.collapse_unifurcations(tree, source=1)
+
+        # If only implicit root remains after pruning dead lineages, error
+        if len(tree.nodes) == 1:
+            raise BirthDeathFitnessError(
+                "All lineages died before stopping condition"
+            )
+
+        return CassiopeiaTree(tree=tree)
+
+    def remove_and_prune_lineage(self, node: int, tree: nx.DiGraph):
+        """Removes a node and prunes the lineage.
+
+        Removes a node and all ancestors of that node that are no longer the
+        ancestor of any leaves. In the context of a lineage tracing
+        experiment, this removes all ancestral nodes that are not the
+        ancestors of any observed samples at the end of the experiment, thus
+        pruning all lineages that died.
+
+        Args:
+            node: The node to be removed
+            tree: The tree to remove the node from
+        """
+        if len(tree.nodes) > 1:
+            curr_parent = list(tree.predecessors(node))[0]
+            tree.remove_node(node)
+            while (
+                tree.out_degree(curr_parent) < 1
+                and tree.in_degree(curr_parent) > 0
+            ):
+                next_parent = list(tree.predecessors(curr_parent))[0]
+                tree.remove_node(curr_parent)
+                curr_parent = next_parent
+
+    def collapse_unifurcations(
+        self, tree: nx.DiGraph, source: int = None
+    ) -> None:
```

Comment:
Same comment as above: Perhaps we can put it into `cassiopeia/data/utilities.py` or something?

Reply:
Yes, I believe this already exists in the `data.utiliites` module, but perhaps only for ete3 trees. Let's add this function there and call it from that library.